ark/utils/_bootstrapping.pyx

```diff
@@ -0,0 +1,276 @@
+        bootstrap_num (int):
+            number of bootstrap iterations
+
+    """
+    cdef DTYPE_t accum
+    cdef Py_ssize_t m1_label, m2_label
+
+    if not m1n or not m2n:
+        return
+
+    for r in range(bootstrap_num):
+        accum = 0
+        _c_permutation(rand_cols, num_choices)
+        for m1_label in pos_labels:
+            for m2_label in rand_cols[:m2n]:
+                accum += dist_mat_bin[m1_label, m2_label]
+        close_num_rand_view[r] = accum
+
+@boundscheck(False) # Deactivate bounds checking
+@wraparound(False)  # Deactivate negative indexing
+@cdivision(True)    # Ignore modulo/divide by zero warning
+cdef inline void _dict_accum(DTYPE_t[:] close_num_rand_view,
+                             const DTYPE_t[:] cols_in_row_flat, const MAXINDEX_t[:] row_indicies,
+                             const MAXINDEX_t[:] pos_labels, Py_ssize_t* rand_cols,
+                             UINT8_t* rand_cols_flags, Py_ssize_t num_choices, int m1n, int m2n,
+                             int bootstrap_num):
+    """ Dictionary based accumulation for large secondary marker size
+
+    Accumulation still loops over a random list of rows, but instead of directly indexing the
+    binarized distance matrix over a random list of columns, the known positive columns within each
+    row are checked against a randomized 'positive column' lookup table (dictionary).
+    
+    This is faster when the average number of positive columns in a row is significantly less than
+    the number of random columns to check (hence why sorting marker column order low to high is
+    encouraged...)
+
+    Args:
+        close_num_rand_view (np.ndarray[np.uint16]):
+            typed memory view of the close_num_rand_view datastructure
+        cols_in_row_flat (np.ndarray[np.uint16]):
+            flattened list-of-lists representation of binarized distance matrix
+        row_indicies (np.ndarray[np.uint64]):
+            'deflattening' index array for `cols_in_row_flat`
+        pos_labels (np.ndarray[np.uint64]):
+            marker indices within dist_mat_bin    
+        rand_cols (Py_ssize_t*):
+            pointer to mutable column randomization memory block
+        rand_cols_flags (uint8_t*):
+            pointer to mutable randomized column lookup table memory block
+        num_choices (Py_ssize_t):
+            number of choices for permutation generation
+        m1n (int):
+            number of rows to select
+        m2n (int):
+            number of columns to select
+        bootstrap_num (int):
+            number of bootstrap iterations
+    """
+    cdef DTYPE_t accum
+    cdef MAXINDEX_t flat_start, flat_end, m2_idx
+    cdef Py_ssize_t m1_label, m2_label
+
+    if not m1n or not m2n:
+        return
+
+    for r in range(bootstrap_num):   
+        accum = 0
+        _c_permutation(rand_cols, num_choices)
+        memset(rand_cols_flags, 0, num_choices * sizeof(UINT8_t))
+        _init_flag_table(rand_cols_flags, rand_cols, m2n)
```

Comment:
This function seems pretty short, can we move the code in `_init_flag_table` here or is this something we expect to see elsewhere in `_bootstrapping.pyx` in the future (or is there something about Cython that makes this advantageous)?

Reply:
I figured that it would look somewhat out of place in that function.  In particular, I think it improves the readability for it the be a separate function.  Plus, I could see this flag-table approach being useful in other calls in the future, since it's basically the closest thing I can have to a python dictionary in cython (hash table, no hash).

Reply:
open to thoughts though